system/jhtree/jhtree.cpp

```diff
@@ -1949,56 +1658,304 @@ bool CKeyCursor::ltEqual(const char *src, char *dst, bool seekForward)
+    if (logExcessiveSeeks && lwildseeks > 1000)
+        reportExcessiveSeeks(lwildseeks, lastSeg, stats);
+    stats.noteSeeks(lseeks, lscans, lwildseeks);
+    return ret;
 }
 
-const byte *CKeyCursor::loadBlob(unsigned __int64 blobid, size32_t &blobsize)
+bool CKeyCursor::lookupSkip(const void *seek, size32_t seekOffset, size32_t seeklen, KeyStatsCollector &stats)
 {
-    const byte *ret = key.loadBlob(blobid, blobsize);
-    activeBlobs.append(ret);
+    if (skipTo(seek, seekOffset, seeklen))
+        stats.noteSkips(1, 0);
+    else
+        stats.noteSkips(0, 1);
+    bool ret = lookup(true, stats);
+#ifdef _DEBUG
+    if (traceSmartStepping)
+    {
+        StringBuffer recstr;
+        unsigned i;
+        for (i = 0; i < key.keySize(); i++)
+        {
+            unsigned char c = ((unsigned char *) keyBuffer)[i];
+            recstr.appendf("%c", isprint(c) ? c : '.');
+        }
+        recstr.append ("    ");
+        for (i = 0; i < key.keySize(); i++)
+        {
+            recstr.appendf("%02x ", ((unsigned char *) keyBuffer)[i]);
+        }
+        DBGLOG("SKIP: Got skips=%02d seeks=%02d scans=%02d : %s", stats.skips, stats.seeks, stats.scans, recstr.str());
+    }
+#endif
     return ret;
 }
 
-void CKeyCursor::releaseBlobs()
+
+unsigned __int64 CKeyCursor::getCount(KeyStatsCollector &stats)
 {
-    ForEachItemIn(idx, activeBlobs)
+    reset();
+    unsigned __int64 result = 0;
+    unsigned lseeks = 0;
+    unsigned lastRealSeg = segs->lastRealSeg();
+    for (;;)
     {
-        free((void *) activeBlobs.item(idx));
+        if (_lookup(true, lastRealSeg, stats))
+        {
+            unsigned __int64 locount = getSequence();
+            endRange(lastRealSeg);
+            ltEqual(keyBuffer, stats);
+            lseeks++;
+            result += getSequence()-locount+1;
+            if (!incrementKey(lastRealSeg))
+                break;
+        }
+        else
+            break;
     }
-    activeBlobs.kill();
+    stats.noteSeeks(lseeks, 0, 0);
+    return result;
 }
 
-bool CKeyCursor::bloomFilterReject(const SegMonitorList &segs) const
+unsigned __int64 CKeyCursor::checkCount(unsigned __int64 max, KeyStatsCollector &stats)
 {
-    return key.bloomFilterReject(segs);
+    setLow(0);
+    reset();
+    unsigned __int64 result = 0;
+    unsigned lseeks = 0;
+    unsigned lastFullSeg = segs->lastFullSeg();
+    if (lastFullSeg == (unsigned) -1)
+    {
+        stats.noteSeeks(1, 0, 0);
+        if (last(nullptr, stats))
+            return getSequence()+1;
+        else
+            return 0;
+    }
+    for (;;)
+    {
+        if (_lookup(true, lastFullSeg, stats))
+        {
+            unsigned __int64 locount = getSequence();
+            endRange(lastFullSeg);
+            ltEqual(keyBuffer, stats);
+            lseeks++;
+            result += getSequence()-locount+1;
+            if (max && (result > max))
+                break;
+            if (!incrementKey(lastFullSeg))
+                break;
+        }
```

Comment:
This used to set matched to false - is the removal deliberate?

Reply:
ltEqual clears it. Whether that is cleaner or not, I don't know!